Implement Scraping Algorithm!

JeremyEastham opened the pull request:
# Additions and Improvements:
- Implemented a **generic** and **simple**, yet **extensible** web scraping algorithm (Closes #8)
- Changed all imports to comply with modern Python Module standards (Closes #24)
- Added `__init__.py` files and renamed `main.py` to `__main__.py` (See #24)
- Added a PyCharm Run Configuration to run `python -m agtern`
- Created scripts to alphabetize `companies.csv` and convert it into the more flexible `scraping_config.json`
- Converted `companies.csv` into `scraping_config.json`
- Added the `agtern.backend` module
  + **Pre-MVP:** This provides functions to initiate web scraping operations in a separate Python process and the results are stored in `db.json`
  + **Post-MVP:** This module will manage a web server to access a database (instead of functions that read a json file)
- Added the `agtern.frontend` module
  + **Pre-MVP:** This provides functions that call `agtern.backend` functions
  + **Post-MVP:** This will provide functions that make web requests to the AgTern server

# To-Do After Merging:
- [ ] Refactor the web scraping algorithm into multiple functions and add comments
- [ ] Discuss the format of `scraping_config.json` and the Internship Model
- [ ] Discuss how the web scraper is dispatched and the parameters that the web scraper needs to start

# Important Notes:
- Due to the changes in imports, the project is now run with `python -m agtern` from the `AgTern` directory (**NOT** `AgTern/agtern`)
- Currently, the Chrome instance controlled by Selenium **NOT** shown, but you can enable it by changing `HEADLESS_SCRAPER` in `__main__.py` to `False` (useful for developer tools)

brendanehrman1 commented on agtern/_companies_csv_to_scraping_config_json.py:
```diff
@@ -0,0 +1,15 @@
+
+import csv, json
+
+data = []
+with open( "companies.csv", "r" ) as f:
+    reader = csv.DictReader( f, fieldnames = [ "company_id", "url" ] )
+    for entry in reader:
+        data.append( {
+            "company": entry["company_id"],
+            "link": entry["url"].strip().strip( "\"" ),
+            "scrape": {}
+        } )
+
+with open( "../scraping_config.json", "w" ) as f:
+    json.dump( data, f, indent = 2 )
```
A couple of things:

1. Is it possible to shorten the name to something like `_process_companies.py` and place the file inside of a `scripts` folder within the project?
2. Instead of overwriting `scraping_config.json` and losing strategies for scraping:

- For every company in companies.csv:
     - Find corresponding company in scraping_config.json
     - If company exists:
         - Update only link
     - If not:
         - Add new object with company/link
         - Set scrape to {} as before

- Write modified scraping_config JSON object to `scraping_config.json`

JeremyEastham replied:
See my comment below for details.

brendanehrman1 commented on agtern/_sort_companies_csv.py:
```diff
@@ -0,0 +1,30 @@
+
+# Read the csv
+with open( "companies.csv", "r" ) as f:
+    data = f.read()
+
+# Map name -> line of csv
+companies = {}
+
+# List of names
+company_names = []
+
+# Load and process csv data
+for line in data.split( "\n" ):
+    if "," not in line:
+        continue
+    company_name = line.split( "," )[0]
+    company_names.append( company_name )
+    companies[company_name] = line
+
+# Sort the names
+company_names = sorted( company_names )
+
+# Construct new csv
```
Is it possible to shorten the name to something like `_sort_companies.py` and place the file inside of a scripts folder within the project (along with `process_companies.py`)?

JeremyEastham replied:
See my comment below for details.

johndpjr commented:
**Good job Jeremy - a few things that @brendanehrman1 and I saw**
- [ ] Add `.idea` folder to the `.gitignore`
- [ ] Move all scripts (`.py` files beginning with only one underscore `_`) into a separate `scripts` folder in `agtern` so that it doesn't pollute `agtern`. Also, try to rename some of the scripts to more concise names (I'm looking at you, `_companies_csv_to_scraping_config_json.py`)
- [ ] Modify the `scrape` function so that you don't overwrite `scraping_config.json` and risk losing strategies for scraping (Tesla, in this example). Instead, you could follow the algorithm below (still keeping everything else)
For every company in `companies.csv`
    If company exists in `scraping_config.json`:
        Update only link (if changed)
    If not:
        Add new object with company & link
        Set scrape to {} as before
        Write modified scraping_config JSON object to `scraping_config.json`
- [ ] `property` in `scraper.py` (line 54) is a reserved keyword in Python, use another name to avoid confusion
- [ ] Make line 62 more readable (I know it loops through the number of internships found, maybe modify or at least add comment describing behavior and why)

**Some notes**
- Formatting should be consistent across the project. We'll let it pass in this merge and create a ticket for enforcing formatting and creating configurations for linting tools so that formatting is consistent
- Should `db.json` and `scraping_config.json` be inside `agtern`? The root dir is more for meta-files (`.gitignore` and so on)

brendanehrman1 commented on agtern/backend/scraper.py:
```diff
@@ -0,0 +1,81 @@
+            driver.get( company_config_entry["link"] )
+
+            data = {}
+            for property in company_config_entry["scrape"].keys():
+                elements = wait.until(
+                        condition.presence_of_all_elements_located(
+                                ( By.XPATH, company_config_entry["scrape"][property] )
+                        )
+                )
+                data[property] = [ element.text for element in elements ]
```
`property` seems to be a reserved keyword in Python. Is it possible to use another name here?

JeremyEastham replied:
I shortened the name to `prop`. It's less descriptive, but it  also won't shadow the `property` function.

brendanehrman1 commented on agtern/backend/scraper.py:
```diff
@@ -0,0 +1,81 @@
+                )
+                data[property] = [ element.text for element in elements ]
+
+            for i in range( len( data[list( data.keys() )[0]] ) ):
```
Is it possible to make this line more readable? Maybe add a variable or comment that describes what this does (like `# iterates through available internships`)?

JeremyEastham replied:
I added a few descriptive comments and a variable that makes this logic clearer.

brendanehrman1 reviewed (changes requested):
Looks good! Just a couple of changes to make - @johndpjr should have some additional feedback at the bottom of his review!

JeremyEastham commented:
@johndpjr and @brendanehrman1, I have re-requested reviews from both of you. I implemented most of your suggestions and added a few features that made sense once I started implementing those changes.

### Change Summary
> Add `.idea` folder to the `.gitignore`

I **did not** take this suggestion because the PyCharm Run Configurations (which are meant to be shared), are located in that directory. If you open `.idea`, you can see that it contains an auto-generated `.gitignore` to ensure that only shared project settings are committed.

> Move all scripts (`.py` files beginning with only one underscore `_`) into a separate `scripts` folder in `agtern` so that it doesn't pollute `agtern`. Also, try to rename some of the scripts to more concise names (I'm looking at you, `_companies_csv_to_scraping_config_json.py`)

I **did** take this suggestion. These scripts were originally intended to be short-lived. I intended to delete them after they served their purpose. However, I realized that they could be useful if there was a standardized way to run them. Thus, I refactored these scripts into a new `agtern.scripts` package. Then, I created the brand new **AgTern CLI**! I created another package, `agtern.cli`, that serves as the new entry point to AgTern. It uses the `argparse` module to parse command line arguments and run the scripts in `agtern.scripts`. This is launched via `python agtern.py`.

```
python agtern.py
Runs the GUI with default settings

python agtern.py --show-scraper
Sets headless to False on ChromeDriver

python agtern.py sort-companies
Runs agtern.scripts.sort_companies
(sorts companies.csv and scraping_config.json)

python agtern.py import-companies
Runs agtern.scripts.import_companies
(imports links from companies.csv into scraping_config.json)
```

> Modify the `scrape` function so that you don't overwrite `scraping_config.json` and risk losing strategies for scraping (Tesla, in this example).

I **did** take this suggestion. I assume that you were talking about the scripts since `agtern.backend.scrape` does not modify the scraping config. In addition, I modified the sorting script to sort both `companies.csv` and `scraping_config.json`.

> Formatting should be consistent across the project. We'll let it pass in this merge and create a ticket for enforcing formatting and creating configurations for linting tools so that formatting is consistent

I agree with this, but we should discuss formatting guidelines as a group before we enforce this.

> * Should `db.json` and `scraping_config.json` be inside `agtern`? The root dir is more for meta-files (`.gitignore` and so on)

I **did** take this suggestion. I moved all of the config and data files into `/data`. I also created the `agtern.data` package that contains a `DataFile` class that standardizes data file creation in the correct location.
```py
from agtern.data import DataFile

# Example:
companies_csv = DataFile( "companies.csv", default_data = "" )
# The above line creates companies.csv in the correct location if it does not exist
# If default_data is not specified, the file isn't automatically created

# Creates the file if it doesn't exist:
companies_csv.create()

# Tests if the file exists:
companies_csv.exists()

# The path of the file can be easily accessed:
with open( companies_csv.path, "r" ) as f:
    ...
```

johndpjr commented on agtern/scripts/import_companies.py:
```diff
@@ -0,0 +1,36 @@
+        for entry in reader:
+            found_existing_entry = False
+            for original_entry in original_config: # Just update link if entry already exists
+                if original_entry["company_id"] == entry["company_id"]:
```
```suggestion
                if original_entry["company"] == entry["company_id"]:
```
Getting a KeyError when running this script - just needed to change `company_id` to company

JeremyEastham replied:
I have updated `import_companies.py` to use `"company"` and `"link"` for the csv columns to match `scraping_config.json`.

johndpjr reviewed (changes requested):
When running `import-companies` script, duplicate companies are appearing in `scraping_config.json`. Please look into this; otherwise, looks really good!

![Screenshot from 2022-10-05 08-05-32](https://user-images.githubusercontent.com/78290887/194067378-73c03f0a-d9f2-4eeb-8730-7caa80391d94.png)

JeremyEastham commented:
Sorry, it looks like I forgot to test `import_companies.py` before I committed my changes. Good catch!

The script is fixed now. This PR is ready to merge!

johndpjr reviewed (approved):
Great job Jeremy; this has made our codebase much more extendable and maintainable. This is a good day for AgTern!

brendanehrman1 reviewed (approved):
Lgtm! Thanks for doing this!!
